lang/csharp/src/apache/main/Util/Decimal.cs

```diff
@@ -0,0 +1,126 @@
+/**
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+using System;
+using Avro.Generic;
+
+namespace Avro.Util
+{
+    /// <summary>
+    /// The 'decimal' logical type.
+    /// </summary>
+    public class Decimal : LogicalType
+    {
+        /// <summary>
+        /// The logical type name for Decimal.
+        /// </summary>
+        public static readonly string LogicalTypeName = "decimal";
+
+        /// <summary>
+        /// Initializes a new Decimal logical type.
+        /// </summary>
+        public Decimal() : base(LogicalTypeName)
+        { }
+
+        /// <summary>
+        /// Applies 'decimal' logical type validation for a given logical schema.
+        /// </summary>
+        /// <param name="schema">The schema to be validated.</param>
+        public override void ValidateSchema(LogicalSchema schema)
+        {
+            if (Schema.Type.Bytes != schema.BaseSchema.Tag && Schema.Type.Fixed != schema.BaseSchema.Tag)
+                throw new AvroTypeException("'decimal' can only be used with an underlying bytes or fixed type");
+        }
+
+        /// <summary>
+        /// Converts a logical value to an instance of its base type.
+        /// </summary>
+        /// <param name="logicalValue">The logical value to convert.</param>
+        /// <param name="schema">The schema that represents the target of the conversion.</param>
+        /// <returns>An object representing the encoded value of the base type.</returns>        
+        public override object ConvertToBaseValue(object logicalValue, LogicalSchema schema)
```

Comment:
@timjroberts I've been diving into avro decimal serialisation in a recent project, and I believe your conversion to/from decimals & bytes doesn't correctly account for 1) endianness and 2) the specific nature of 'valBits[3]' in the decimal type's data structure (according to the docs https://docs.microsoft.com/en-us/dotnet/api/system.decimal it contains a sign flag & the scale).

I've implemented some conversion code here: https://github.com/petersilverwood/avro-decimal-tools/tree/master/lang/csharp with some test fixtures to verify the same serialisation in C# and in java. There are probably some edge cases to handle and I haven't tested performance, but it should be good starting point.

Is this useful to integrate into your PR?

Reply:
Yes, the implementations of the logical types were naive and likely not fully meeting with the current Avro specification. It was the one area where I needed more focused review. I'll take a look at your referenced. Many thanks.

Reply:
i have some scripts in a local branch for bringing up an integration test environment in docker (multi broker kafka cluster, schema registry, monitoring etc). once that is in, it will be easy to get java <-> .net integration tests going. i have been thinking that is a pre-requisite for getting this PR merged. this is somewhere high on my priority list, but not at the top.